### ISSUE TITLE
Fuzzer fails with "Found minimal failing expression"

### ISSUE THREAD

**Issue by ashokku2022** [facebookincubator/velox#4401]
### Description

Fuzzer test should pass, however, seeing fuzzer hitting runtime error. I see this on my PR as well on the other too.

https://github.com/facebookincubator/velox/pull/4364
https://github.com/facebookincubator/velox/pull/4394

```
E20230323 18:07:34.657228 231274 Exceptions.h:68] Line: ../../velox/expression/tests/ExpressionFuzzer.cpp:1176, Function:errorExit, Expression:  Found minimal failing expression, Source: RUNTIME, ErrorCode: INVALID_STATE
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Found minimal failing expression
Retriable: False
Function: errorExit
File: ../../velox/expression/tests/ExpressionFuzzer.cpp
Line: 1176
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxException5State4makeIZNS1_C4EPKcmS5_St17basic_string_viewIcSt11char_traitsIcEES9_S9_S9_bNS1_4TypeES9_EUlRT_E_EESt10shared_ptrIKS2_ESA_SB_
# 2  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 3  _ZN8facebook5velox17VeloxRuntimeErrorC2EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bS7_
# 4  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
# 5  _ZN8facebook5velox4test16ExpressionFuzzer9errorExitERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
# 6  _ZN8facebook5velox4test16ExpressionFuzzer9drilldownESt10shared_ptrIKNS0_4core10ITypedExprEERKS3_INS0_9RowVectorEERKSt6vectorIjSaIjEE
# 7  _ZN8facebook5velox4test16ExpressionFuzzer2goEv
# 8  _ZN8facebook5velox4test16expressionFuzzerESt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIPKNS0_4exec17FunctionSignatureESaISD_EESt4hashIS8_ESt8equal_toIS8_ESaISt4pairIKS8_SF_EEEm
# 9  _ZN12FuzzerRunner3runERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmRKSt13unordered_setIS5_St4hashIS5_ESt8equal_toIS5_ESaIS5_EES7_
# 10 main
# 11 __libc_start_main
# 12 _start

*** Aborted at 1679594854 (Unix time, try 'date -d @1679594854') ***
*** Signal 6 (SIGABRT) (0x3876a) received by PID 231274 (pthread TID 0x7f7ef565a2c0) (linux TID 231274) (maybe from PID 231274, UID 0) (code: -6), stack trace: ***
(error retrieving stack trace)
/bin/bash: line 14: 231274 Aborted                 (core dumped) _build/debug/velox/expression/tests/velox_expression_fuzzer_test --seed ${RANDOM} --enable_variadic_signatures --velox_fuzzer_enable_complex_types --lazy_vector_generation_ratio 0.2 --velox_fuzzer_enable_column_reuse --velox_fuzzer_enable_expression_reuse --retry_with_try --duration_sec 60 --logtostderr=1 --minloglevel=0 --repro_persist_path=/tmp/fuzzer_repro

Exited with code exit status 134
CircleCI received exit code 134
```

### Error Reproduction

`(core dumped) _build/debug/velox/expression/tests/velox_expression_fuzzer_test --seed ${RANDOM} --enable_variadic_signatures --velox_fuzzer_enable_complex_types --lazy_vector_generation_ratio 0.2 --velox_fuzzer_enable_column_reuse --velox_fuzzer_enable_expression_reuse --retry_with_try --duration_sec 60 --logtostderr=1 --minloglevel=0 --repro_persist_path=/tmp/fuzzer_repro`
``

### Relevant logs

```bash
E20230323 18:07:34.657228 231274 Exceptions.h:68] Line: ../../velox/expression/tests/ExpressionFuzzer.cpp:1176, Function:errorExit, Expression:  Found minimal failing expression, Source: RUNTIME, ErrorCode: INVALID_STATE
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Found minimal failing expression
Retriable: False
Function: errorExit
File: ../../velox/expression/tests/ExpressionFuzzer.cpp
Line: 1176
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxException5State4makeIZNS1_C4EPKcmS5_St17basic_string_viewIcSt11char_traitsIcEES9_S9_S9_bNS1_4TypeES9_EUlRT_E_EESt10shared_ptrIKS2_ESA_SB_
# 2  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 3  _ZN8facebook5velox17VeloxRuntimeErrorC2EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bS7_
# 4  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
# 5  _ZN8facebook5velox4test16ExpressionFuzzer9errorExitERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
# 6  _ZN8facebook5velox4test16ExpressionFuzzer9drilldownESt10shared_ptrIKNS0_4core10ITypedExprEERKS3_INS0_9RowVectorEERKSt6vectorIjSaIjEE
# 7  _ZN8facebook5velox4test16ExpressionFuzzer2goEv
# 8  _ZN8facebook5velox4test16expressionFuzzerESt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIPKNS0_4exec17FunctionSignatureESaISD_EESt4hashIS8_ESt8equal_toIS8_ESaISt4pairIKS8_SF_EEEm
# 9  _ZN12FuzzerRunner3runERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmRKSt13unordered_setIS5_St4hashIS5_ESt8equal_toIS5_ESaIS5_EES7_
# 10 main
# 11 __libc_start_main
# 12 _start
```

**Comment by kgpai** [comment]
I will take a look soon.

**Comment by oerling** [comment]
If an expression fails, it tries smaller expressions until it find a failing expressions of which all child expressions pass. It then reports this as hte minimal failing expr. this is a feature. So, to debug the minimal failing expression, put a breakpoint at the line where it says 'breakpoint here' and debug the expression. If there were no failing expression to begin with and this path was invoked then this would be an error. So, is it the case that there is no fuzzer failure and it still exits this way? Note that the original behavior was to ie with unhandled exception on first failure. So, I continued this practice also here. 

Note that if you are debugging special forms, the logic for applying the test to successively smaller expressions might have to be customized.

**Comment by kgpai** [comment]
Hi Ashok, 

I think you have a couple of different issues: 

On the first diff https://github.com/facebookincubator/velox/pull/4364 : 
 What you are seeing is a failure occurring for an expression on rows that shouldn't fail. This is a valid failure that should be investigated.  I have raised an issue and will investigate soon : https://github.com/facebookincubator/velox/issues/4247 . (It seems this problem is more common than I anticipated). 
 
 On the 2nd diff https://github.com/facebookincubator/velox/pull/4394 : 
Here the fuzzer runs into the issue below and tries to find the smallest possible expression that causes this problem. In this case the expression `map_concat(try_cast <empty> as MAP<VARCHAR,BIGINT>..` cant be simplified any further and it finally says that it found the minimal expression along with the root cause of the problem it ran into. 
Note in this case the exception is coming from the fuzzer verifier and must be investigated. 
I dont think an issue exists for problem below and I am going to open one for it. 


```
E20230323 18:07:34.657006 231274 Exceptions.h:68] Line: ../.././velox/vector/FlatVector-inl.h:229, Function:copyValuesAndNulls, Expression: source->size() >= sourceIndex + count (0 vs. 713), Source: RUNTIME, ErrorCode: INVALID_STATE
E20230323 18:07:34.657228 231274 Exceptions.h:68] Line: ../../velox/expression/tests/ExpressionFuzzer.cpp:1176, Function:errorExit, Expression:  Found minimal failing expression, Source: RUNTIME, ErrorCode: INVALID_STATE
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Found minimal failing expression
Retriable: False
Function: errorExit
File: ../../velox/expression/tests/ExpressionFuzzer.cpp
Line: 1176
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxException5State4makeIZNS1_C4EPKcmS5_St17basic_string_viewIcSt11char_traitsIcEES9_S9_S9_bNS1_4TypeES9_EUlRT_E_EESt10shared_ptrIKS2_ESA_SB_
# 2  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 3  _ZN8facebook5velox17VeloxRuntimeErrorC2EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bS7_
# 4  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
# 5  _ZN8facebook5velox4test16ExpressionFuzzer9errorExitERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
# 6  _ZN8facebook5velox4test16ExpressionFuzzer9drilldownESt10shared_ptrIKNS0_4core10ITypedExprEERKS3_INS0_9RowVectorEERKSt6vectorIjSaIjEE
# 7  _ZN8facebook5velox4test16ExpressionFuzzer2goEv
# 8  _ZN8facebook5velox4test16expressionFuzzerESt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIPKNS0_4exec17FunctionSignatureESaISD_EESt4hashIS8_ESt8equal_toIS8_ESaISt4pairIKS8_SF_EEEm
# 9  _ZN12FuzzerRunner3runERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmRKSt13unordered_setIS5_St4hashIS5_ESt8equal_toIS5_ESaIS5_EES7_
# 10 main
# 11 __libc_start_main
# 12 _start```

FYI, None of these issues seem related to your code and I dont think your PR should be gated on these fixes ; While we should do a better job in communicating these failures from fuzzer at the moment this behavior does not look like bugs to me.

**Comment by ashokku2022** [comment]
Thanks Krishna, appreciate your response!

**Comment by kgpai** [comment]
@pedroerp I will create a PR right now to revert the minimal expression change and a future PR will add this flag that allows us to drill down and get a smaller repro during debugging.

**Comment by pedroerp** [comment]
Maybe we don't need to revert it, but rather just hide it behind a flag?

**Comment by kgpai** [comment]
@pedroerp Saw this too late; Already have a revert Pr : https://github.com/facebookincubator/velox/pull/4409 which is landing.  I think we still need to have better messaging , so probably best to have another pr with the flag and messaging rather than just a flag. I will re-enable this soon with all of the above though.
